icu4c/source/common/locid.cpp

```diff
@@ -504,34 +508,1071 @@ Locale::operator==( const   Locale& other) const
+        UErrorCode &status)
+{
+    if (U_FAILURE(status)) {
+        return;
+    }
+    length = ures_getSize(alias);
+
+    const char** rawTypes = types.allocateInsteadAndCopy(length);
+    int32_t* rawIndexes = replacementIndexes.allocateInsteadAndCopy(length);
+    int i = 0;
+    while (ures_hasNext(alias)) {
+        LocalUResourceBundlePointer res(
+            ures_getNextResource(alias, nullptr, &status));
+        const char *aliasFrom = ures_getKey(res.getAlias());
+        U_ASSERT(uprv_strlen(aliasFrom) >= 4 && uprv_strlen(aliasFrom) <= 8);
+        U_ASSERT(uprv_strlen(aliasFrom) != 4 ||
+                 (aliasFrom[0] >= 'a' && aliasFrom[0] <= 'z'));
+        rawTypes[i] = aliasFrom;
+        UnicodeString aliasTo =
+            ures_getUnicodeStringByKey(res.getAlias(), "replacement", &status);
+        U_ASSERT(aliasTo.length() >= 4 && aliasTo.length() <= 8);
+        U_ASSERT(aliasTo.length() != 4 ||
+                 (aliasTo.charAt(0) >= 'a' && aliasTo.charAt(0) <= 'z'));
+        rawIndexes[i] = strings->add(aliasTo, status);
+        i++;
+    }
+}
+
+/**
+ * Initializes the alias data from the ICU resource bundles. The alias data
+ * contains alias of language, country, script and variants.
+ *
+ * If the alias data has already loaded, then this method simply returns without
+ * doing anything meaningful.
+ */
+void U_CALLCONV
+AliasData::loadData(UErrorCode &status)
+{
+#ifdef LOCALE_CANONICALIZATION_DEBUG
+    UDate start = uprv_getRawUTCtime();
+#endif  // LOCALE_CANONICALIZATION_DEBUG
+    ucln_common_registerCleanup(UCLN_COMMON_LOCALE_ALIAS, cleanup);
+    AliasDataBuilder builder;
+    gSingleton = builder.build(status);
+#ifdef LOCALE_CANONICALIZATION_DEBUG
+    UDate end = uprv_getRawUTCtime();
+    printf("AliasData::loadData took total %f ms\n", end - start);
+#endif  // LOCALE_CANONICALIZATION_DEBUG
+}
+
+/**
+ * Build the alias data from resources.
+ */
+AliasData*
+AliasDataBuilder::build(UErrorCode &status) {
+    LocalUResourceBundlePointer metadata(
+        ures_openDirect(nullptr, "metadata", &status));
+    LocalUResourceBundlePointer metadataAlias(
+        ures_getByKey(metadata.getAlias(), "alias", nullptr, &status));
+    LocalUResourceBundlePointer languageAlias(
+        ures_getByKey(metadataAlias.getAlias(), "language", nullptr, &status));
+    LocalUResourceBundlePointer scriptAlias(
+        ures_getByKey(metadataAlias.getAlias(), "script", nullptr, &status));
+    LocalUResourceBundlePointer territoryAlias(
+        ures_getByKey(metadataAlias.getAlias(), "territory", nullptr, &status));
+    LocalUResourceBundlePointer variantAlias(
+        ures_getByKey(metadataAlias.getAlias(), "variant", nullptr, &status));
+
+    int32_t languagesLength = 0, scriptLength = 0, territoryLength = 0,
+            variantLength = 0;
+
+    // Read the languageAlias into languageTypes, languageReplacementIndexes
+    // and strings
+    LocalPointer<UniqueCharStrings> strings(new UniqueCharStrings(status));
```

Comment:
This is where global "new" is used because UniqueCharStrings does not extend UMemory. Please stack-allocate the "strings" object, not heap-allocate it.

Reply:
ok I will do it, but it is just a shock we cannot use new in C++.

Reply:
We can and do use operator new on subclasses of UMemory. If needed, we can make UniqueCharStrings be such a subclass, but I don't think we need it.

https://unicode-org.github.io/icu/userguide/dev/codingguidelines.html#memory-allocation